src/librustc_resolve/error_reporting.rs

```diff
@@ -564,10 +565,290 @@ impl<'a, 'b:'a> ImportResolver<'a, 'b> {
+                    //   ie. `use a::b::{c, d};`
+                    //                      ^
+                    // Then expand the span of characters to remove to include the previous
+                    // binding's trailing comma.
+                    //   ie. `use a::b::{c, d};`
+                    //                    ^^^
+                    if let Some(previous_span) = extend_span_to_previous_binding(
+                        self.resolver.session, binding_span,
+                    ) {
+                        debug!("check_for_module_export_macro: previous_span={:?}", previous_span);
+                        removal_span = removal_span.with_lo(previous_span.lo());
+                    }
+                }
+                debug!("check_for_module_export_macro: removal_span={:?}", removal_span);
+
+                // Find the span after the crate name and if it has nested imports immediatately
+                // after the crate name already.
+                //   ie. `use a::b::{c, d};`
+                //               ^^^^^^^^^
+                //   or  `use a::{b, c, d}};`
+                //               ^^^^^^^^^^^
+                let (has_nested, after_crate_name) = find_span_immediately_after_crate_name(
+                    self.resolver.session, module_name, directive.use_span,
+                );
+                debug!("check_for_module_export_macro: has_nested={:?} after_crate_name={:?}",
+                       has_nested, after_crate_name);
+
+                let source_map = self.resolver.session.source_map();
+
+                // Remove two bytes at the end to keep all but the `};` characters.
+                //   ie. `{b::{c, d}, e::{f, g}};`
+                //        ^^^^^^^^^^^^^^^^^^^^^
+                let end_bytes = BytePos(if has_nested { 2 } else { 1 });
+                let mut remaining_span = after_crate_name.with_hi(
+                    after_crate_name.hi() - end_bytes);
+                if has_nested {
+                    // Remove two bytes at the start to keep all but the initial `{` character.
+                    //   ie. `{b::{c, d}, e::{f, g}`
+                    //         ^^^^^^^^^^^^^^^^^^^^
+                    remaining_span = remaining_span.with_lo(after_crate_name.lo() + BytePos(1));
+                }
+
+                // Calculate the number of characters into a snippet to remove the removal
+                // span.
+                let lo = removal_span.lo() - remaining_span.lo();
+                let hi = lo + (removal_span.hi() - removal_span.lo());
+                if let Ok(mut remaining) = source_map.span_to_snippet(remaining_span) {
+                    // Remove the original location of the binding.
+                    remaining.replace_range((lo.0 as usize)..(hi.0 as usize), "");
+                    correction = format!("use {}::{{{}, {}}};", module_name, import, remaining);
```

Comment:
This has the potential to mangle the formatting severely, so I'd prefer a multipart suggestion for removing and adding the imports in reasonable places, instead of trying to synthesize a single span suggestion. That would still have problems though, namely having the moved imports all in the first line and keeping an empty line with all its whitespace, but I feel that we could address the later issue from the suggestion engine itself. The comments are great, but the span tweaking itself gives me pause, as I've introduced bugs doing similar things in the past.

Reply:
I've changed this to a multipart suggestion.